Build: enhance IDL test harness by adding write tests and more read tests

mwinkel-dev opened the pull request:
Issue #2650

This is the first draft of the enhanced IDL test harness.   (It will surely be revised before being upgraded to a PR.)

Note that the write tests presently use a local tree on the build server.   After some infrastructure upgrades, the write tests will be changed to use a mdsip connection to write to a tree on a different server.

Comments on this work in progress will be appreciated.

ModestMC commented on idl/testing/run_tests.py:
```diff
@@ -1213,6 +1325,119 @@ def idl_test(code, expected_output):
+mdsput, 'SUBTREE_1.E_UNITS', 'build_with_units(55, "volts")'
+mdsset_def, 'SUBTREE_1'
+mdsput, '.-.SUBTREE_2:F_NUM', '66'
+mdsput, '\TAG_G', '77'
```
Do you ever try reading the data from this tag (or just some tag)? I looked and I don't think I saw something but I could have missed it.

ModestMC replied:
I believe in line 1412 you read the data stored here, but it's not referenced by the tag specifically.

mwinkel-dev replied:
Hi @ModestMC -- In the `IDL-read-various` function at line 1344 is a read using a tag.

You are correct that in the `IDL-write-various` function the value written by `\TAG_G` is checked at line 1412 to confirm that the write by tag succeeded.

mwinkel-dev replied:
The line numbers changed significantly when switched to the `argparse` template.   The `IDL-read-various` read of a tag is now around line 1518.   And the `IDL-write-various` check of a tag is around line 1586.

ModestMC commented:
Overall, I think this is really solid progress and I'm personally pretty happy with it. I think there are some more cases to add, but this covers a lot of the key stuff. 

One thing I think would be smart/helpful is maybe some kind of comment block at the top that lists the particular features which are known to be tested and whatever keyword(s) will mostly get a ctrl+f to find the lines of the code relevant for that test. I am definitely going to forget which things get tested where in this file and I figure such a thing will be useful when you guys want to rip out pieces to stick in the upgraded testing framework later.

Finally, I will also note that most of our users interact with Atlas via mdsip, so we're very glad it's on you guys' radar and we'll be keeping an eye out for it.

mwinkel-dev commented:
Thanks for reviewing the enhanced IDL tests.   Goal of this round is to cover the basic features of the API.   After we get the additional infrastructure set up, we'll beef up the mdsip testing.   And eventually, could add even more coverage of the IDL API if needed.

Commit:
Build: enhanced IDL tests now use argparse

mwinkel-dev commented:
Retest this please

mwinkel-dev commented:
retest this please

mwinkel-dev commented:
Retest this please

ModestMC commented:
Not sure if it'll matter, but you may want to add some tests that check for case sensitivity (depending how IDL deals with string datatypes and what magic characters it cares about).

Commit:
Build: parameterize the "write" tree for IDL tests

- no longer uses hard coded tree name and shot number
- also changed file permissions on run_tests.py

WhoBrokeTheBuild commented on idl/testing/run_tests.py:
```diff
@@ -58,8 +58,227 @@
+    import os
+    import MDSplus as mds
+
+    test_dir = os.getenv('MDSPLUS_DIR') + '/idl/testing/'
```
We should instead use a `tempfile.TemporaryDirectory`, but be sure to call that once at the top, not once for each test

mwinkel-dev replied:
That is a great suggestion!

mwinkel-dev replied:
The `/idl/testing` directory is where we already create the transient `test.pro` files that `run_tests.py` executes.   Should those `test.pro` files also be created in a `tempfile.TemporaryDirectory`?

The local tree used for testing the write features of the IDL API is only ~3 KB in size.   Which is why I placed it in the same `idl/testing` directory as the transient `test.pro` files.

Regardless of which directory is used, I favor having the transient `test.pro` files and the dinky local "write" tree in the same directory.  But am open to using two different directories if that makes Jenkins more robust.

Also note that after we upgrade the infrastructure to do mdsip testing and cross-version testing, instead of a local tree we instead need to create the dinky "write" tree on the remote server. 

Regardless, I am now experimenting with `tempfile.TemporaryDirectory`.   But will hold off on making that commit until we decide whether to move the transient `test.pro` files too.

mwinkel-dev replied:
A nice feature of using `tempfile.TemporaryDirectory` is that the directory will be empty on creation and will be deleted after the test completes.   This means that a few lines could be deleted from the existing code.

The same feature though also has a downside.   If the "write" test fails, might be nice to have the dinky tree left behind so that we can inspect it and figure out what just broke in MDSplus.

WhoBrokeTheBuild replied:
Actually, yeah we should probably put the `.pro` there, it'll keep a truly temporary file out of the working directories

mwinkel-dev replied:
A-OK, I will move the `test.pro` file to the `tempfile.TemporaryDirectory` too.

mwinkel-dev commented:
Hi @ModestMC,

Thanks for suggesting that I also add some tests for lowercase items (tags, node names, node paths).   That is a good idea.
https://github.com/MDSplus/mdsplus/pull/2656#issuecomment-1939788672

The goal is to gradually evolve this test into the mdsip / cross-version testing of the API for IDL.   That evolution will take several iterations.  My plan is to get this version merged in now as that will add tests of using the API to write to a tree.   In the next iteration, I will add some lowercase items.

Commit:
Build: use temporary directory for transient files
